x-pack/examples/embedded_lens_example/public/app.tsx

```diff
@@ -113,19 +91,56 @@ function getLensAttributes(
 export const App = (props: {
   core: CoreStart;
   plugins: StartDependencies;
-  defaultIndexPattern: IndexPattern | null;
+  defaultDataView: DataView | null;
 }) => {
   const [color, setColor] = useState('green');
+  const [dataLayer, setDataLayer] = useState<PersistedIndexPatternLayer>();
   const [isLoading, setIsLoading] = useState(false);
   const [isSaveModalVisible, setIsSaveModalVisible] = useState(false);
   const LensComponent = props.plugins.lens.EmbeddableComponent;
   const LensSaveModalComponent = props.plugins.lens.SaveModalComponent;
 
+  useEffect(() => {
+    const updateDataLayer = async (dataView: DataView) => {
```

Comment:
Looking at how the helper is used makes me wonder whether it's a bit cumbersome for the consumer to deal with the async-ness. I see why it is async though, it makes total sense.

Maybe we could go with a structure like this:
```
// Called outside of react when setting up the app
const helper = await props.plugins.lens.formula.createFormulaHelper(dataView.id);
// ...

// Generate layer config synchronously when actually needed
const formulaLayer = helper.insertOrReplaceFormulaColumn(...);
```

What do you think @alexwizp @shahzad31 @dej611 ?

Reply:
@flash1293 in that case we probably need to load `LensIndexPattern` outside of `insertOrReplaceFormulaColumn` method. Not sure we want to make `Lens` version of `loadIndexPatterns` public

I'm talking about next code:
 ```
const indexPatterns = await loadIndexPatterns({
            cache: {},
            patterns: [dataViewId],
            indexPatternsService: startDependencies.data.indexPatterns,
          });
```

Reply:
What I meant is replacing the existing `insertOrReplaceFormulaColumn` on the contract with this:
```ts
createFormulaHelper: (dataViewId: string) => {
     const { loadIndexPatterns, insertOrReplaceFormulaColumn, operationDefinitionMap } =
            await import('./async_services');

          const indexPatterns = await loadIndexPatterns({
            cache: {},
            patterns: [dataViewId],
            indexPatternsService: startDependencies.data.indexPatterns,
          });

         return  (id, { formula, label }, layer, params) => {
          return indexPatterns[dataViewId]
            ? insertOrReplaceFormulaColumn(
                id,
                {
                  label: label ?? formula,
                  customLabel: Boolean(label),
                  operationType: 'formula',
                  dataType: 'number',
                  references: [],
                  isBucketed: false,
                  params: {
                    formula,
                  },
                },
                { ...layer, indexPatternId: dataViewId },
                {
                  indexPattern: indexPatterns[dataViewId],
                  operations:
                    params?.operations?.reduce(
                      (operationsMap, item) => ({
                        ...operationsMap,
                        [item]: operationDefinitionMap[item],
                      }),
                      {}
                    ) ?? undefined,
                }
              ).layer
            : undefined;
   };
}
```

Reply:
Maybe we should mark it as `experimental` and when we do refactoring Lens to use the DataView we can probably revisit that helper.

Reply:
The only thing I do not like in the proposal is the unused cache here:

```ts
const indexPatterns = await loadIndexPatterns({
            cache: {},
            patterns: [dataViewId],
            indexPatternsService: startDependencies.data.indexPatterns,
          });
```

Perhaps the user will most probably use the same indexpattern(s) so the amount of helpers will be fixed, so the unused cache won't matter.
We might expose a wrapper for `loadIndexPatterns` but then it would be non-trivial to explain to the consumer why to use that instead of passing a regular kibana indexPattern/dataView .

All that said, I think I'm leaning on:
* have an explicit name for the factory: `createFormulaHelperForDataView( dataViewId: string )`
* Mark the API as `experimental` and record some feedback after some usage

Reply:
@dej611 Discussed this offline with Alexey and we came up with a nice approach that has a bunch of upsides - he will update the PR